### PR TITLE
Measurement delay compensation

### PR DESCRIPTION
**Pull request by Flova**
## Proposed changes
- Apply odom difference to the filter state as the filter runs slightly delayed due to the image processing.
- Also fix small todo related to params

### CODE REVIEW

File: bitbots_localization/src/localization.cpp [hunk 1]
```diff
@@ -314,6 +314,8 @@ void Localization::updateMeasurements() {
   last_stamp_lines_pc = line_pointcloud_relative_.header.stamp;
   last_stamp_goals = goal_posts_relative_.header.stamp;
   last_stamp_fb_points = fieldboundary_relative_.header.stamp;
+  // Maximum time stamp of the last measurements
+  last_stamp_all_measurments = std::max(std::max(last_stamp_lines_pc, last_stamp_goals), last_stamp_fb_points);
```
Comment (timonegk) [inline bitbots_localization/src/localization.cpp:318]: I think you can use the following:

```suggestion
  last_stamp_all_measurments = std::max({last_stamp_lines_pc, last_stamp_goals, last_stamp_fb_points});
```

File: bitbots_localization/src/localization.cpp [hunk 2]
```diff
@@ -361,50 +363,77 @@ void Localization::getMotion() {
...
+  tf2::Transform filter_transform;
+  filter_transform.setOrigin(tf2::Vector3(estimate.getXPos() , estimate.getYPos(), 0.0));
   tf2::Quaternion q;
-  q.setRPY(0, 0, estimate_.getTheta());
-  q.normalize();
+  q.setRPY(0, 0, estimate.getTheta());
+  filter_transform.setRotation(q);
+
+  // Get the transform from the last measurement timestamp until now
+  geometry_msgs::msg::TransformStamped odomDuringMeasurement, odomNow;
```
Comment (timonegk) [inline bitbots_localization/src/localization.cpp:376]: All of the other variables are in snake_case, so I would prefer to do the same here.

**Comment by jaagut** [comment]
When testing, this change does not break anything, but it is hard to observe if it improves anything.

@Flova maybe you could add a little motivation for this change to this PR description.
Also, how can we test, if it improves things?

**Comment by Flova** [comment]
The filter judges particles based on the observations. But our observations are delayed (vision delay, ipm, etc .) so the filter state is also delayed. Ideally we would keep track of the history of every particle and match the observation with the particle state at that time, but this is not trivial to implement and resource intensive. This is more of an approximate solution where we apply the odometry movement that happened during the measurement delay to our filter state before publishing it. It works well for single modality observations (e.g. camera inputs). It should solve the issue of our localization lagging slightly behind the real robot position, leading to e.g. overshooting in the path planning etc.. It should also allow us to walk faster as the error is correlated to the walking speed.

File: bitbots_localization/src/localization.cpp (hunk 2 above)
Comment (jaagut) [reply bitbots_localization/src/localization.cpp:376]: Other transforms are also in camel case.
Our future linter will complain soon(TM)

**Review by jaagut — APPROVED** [review]

**Pull request merged**